src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+        }
+    }
+
+    /// Given two specs, returns a new spec that only matches things that both original specs would match.
+    pub fn intersect(self, other: ResultSpec<'a>) -> ResultSpec<'a> {
+        ResultSpec {
+            nullable: self.nullable && other.nullable,
+            fallible: self.fallible && other.fallible,
+            values: self.values.intersect(other.values),
+        }
+    }
+
+    /// Check if a particular value matches the spec.
+    pub fn may_contain(&self, value: Datum<'a>) -> bool {
+        if value == Datum::Null {
+            return self.nullable;
+        }
+
+        self.values.may_contain(value)
+    }
+
+    /// Check if an error value matches the spec.
+    pub fn may_fail(&self) -> bool {
+        self.fallible
+    }
+
+    /// This function "maps" a function across the set of valid datums, returning the set of possible
+    /// results.
+    ///
+    /// If we actually stored a set of every possible value, this could by done by passing
+    /// each datup to the function one-by-one and unioning the resulting sets. This is possible
+    /// when our values set is empty or contains a single datum, but when it contains a range,
+    /// we need a little metadata about the function to be able to infer a useful result.
+    fn flat_map(
+        self,
+        is_monotonic: Monotonic,
+        mut datum_map: impl FnMut(Datum<'a>) -> ResultSpec<'a>,
+    ) -> ResultSpec<'a> {
+        let null_spec = if self.nullable {
+            datum_map(Datum::Null)
+        } else {
+            ResultSpec::nothing()
+        };
+
+        let error_spec = ResultSpec {
+            fallible: self.fallible,
+            ..Self::nothing()
+        };
+
+        let values_spec = match self.values {
+            Values::Empty => ResultSpec::nothing(),
+            // If this range contains a single datum, just call the function.
+            Values::Within(min, max) if min == max => datum_map(min),
+            // If this is a range of booleans, we know all the values... just try them.
+            Values::Within(Datum::False, Datum::True) => {
+                datum_map(Datum::False).union(datum_map(Datum::True))
+            }
+            // Otherwise, if our function is monotonic, we can try mapping the input
+            // range to an output range.
+            Values::Within(min, max) => match is_monotonic {
+                Monotonic::Yes => {
+                    let min_column = datum_map(min);
+                    let max_column = datum_map(max);
+                    if min_column.nullable
+                        || min_column.fallible
+                        || max_column.nullable
+                        || max_column.fallible
+                    {
+                        ResultSpec::anything()
+                    } else {
+                        min_column.union(max_column)
+                    }
+                }
+                Monotonic::Maybe | Monotonic::No => ResultSpec::anything(),
```

Comment:
We can narrow down the `nullability` and `fallibility` of the `ResultSpec` if we know more about the function, e.g., if the output type is non-nullability, we can propagate that information.

Reply:
`*Func::output_type(input_cols)` will do an okay job at ruling out some nulls, so I guess we've already got an okay baseline for that.